eng/pipelines/templates/jobs/archetype-sdk-client.yml

```diff
@@ -281,9 +281,43 @@ jobs:
           jdkArchitectureOption: 'x64'
           publishJUnitResults: false
           goals: ${{ parameters.TestGoals }}
-        # we want to run this when TestFromSource isn't true (which covers normal running when it isn't set)
-        # OR when ShouldRunSourceTests is true
-        condition: and(succeeded(), or(ne(variables['TestFromSource'],'true'), eq(variables['ShouldRunSourceTests'],'true')))
+        # we want to run this when TestFromSource isn't true
+        condition: and(succeeded(), ne(variables['TestFromSource'],'true'))
+
+      - task: Maven@3
+        displayName: 'Build for From Source run'
+        inputs:
+          mavenPomFile: pom.xml
+          goals: 'install'
+          # Build everything in the root pom.xml and install it. The purpose here is to build and install everything after
+          # the version have been manipulated. The purpose is to have all of the libraries installed on the machine so we
+          # can simply run tests with the -amd switch and it won't have to deal with the AdditionalModules snafu that's causing
+          # us to list every transitive azure client dependency.
+          # Note: yes, this overbuilds, it builds data track libraries but that'll be looked at later.
+          options: '$(DefaultOptions) -DskipTests -Dgpg.skip -Dmaven.javadoc.skip=true -Dcheckstyle.skip=true -Dspotbugs.skip=true -Drevapi.skip=true'
+          mavenOptions: '$(MemoryOptions) $(LoggingOptions)'
+          javaHomeOption: 'JDKVersion'
+          jdkVersionOption: $(JavaBuildVersion)
+          jdkArchitectureOption: 'x64'
+          publishJUnitResults: false
+        # We want to run this when we TestFromSource and ShouldRunSourceTests are true
+        condition: and(succeeded(), and(eq(variables['TestFromSource'],'true'), eq(variables['ShouldRunSourceTests'],'true')))
+
+      - task: Maven@3
+        displayName: 'Run tests for From Source run'
+        inputs:
+          mavenPomFile: pom.xml
+          # For the From Source runs we don't want the -am switch as we don't care about running tests for our dependencies
+          # but we do want the -amd switch because we want to run tests on things that depend on us.
+          options: ${{ parameters.TestOptions }} -pl $(PLSkipOptions)$(ProjectList) -amd
+          mavenOptions: '$(MemoryOptions) $(LoggingOptions)'
+          javaHomeOption: 'JDKVersion'
+          jdkVersionOption: $(JavaTestVersion)
+          jdkArchitectureOption: 'x64'
+          publishJUnitResults: false
+          goals: ${{ parameters.TestGoals }}
```

Comment:
Given that the step above will install the entire repository this goal could be changed to prevent compile and test compile from happening again, as they trigger when running `test` or `verify`.

`jacoco:prepare-agent jacoco:instrument surefire:test jacoco:restore-instrumented-classes jacoco:report` would effectively replicate everything we want to do in testing. This will setup the Jacoco code coverage, run the tests and instrument code coverage, and then report the coverage data and restore the installed code back to its pre-instrumented state (Jacoco needs to do some byte code re-writing to get coverage).

Doing this would likely save a few minutes in the step.

Reply:
As per our conversation, the next commit will have this to try them out.